pkg/kubelet/cm/dra/plugin/client.go

```diff
@@ -0,0 +1,231 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package dra
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/credentials/insecure"
+	"google.golang.org/grpc/status"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/klog/v2"
+
+	drapbv1 "k8s.io/kubelet/pkg/apis/dra/v1alpha1"
+)
+
+type Client interface {
+	NodePrepareResource(
+		ctx context.Context,
+		namespace string,
+		claimUID types.UID,
+		claimName string,
+		resourceHandle string,
+	) (*drapbv1.NodePrepareResourceResponse, error)
+
+	NodeUnprepareResource(
+		ctx context.Context,
+		namespace string,
+		claimUID types.UID,
+		claimName string,
+		cdiDevice []string,
+	) (*drapbv1.NodeUnprepareResourceResponse, error)
+}
+
+// Strongly typed address.
+type draAddr string
+
+// draPluginClient encapsulates all dra plugin methods.
+type draPluginClient struct {
+	pluginName          string
+	addr                draAddr
+	nodeV1ClientCreator nodeV1ClientCreator
+}
+
+var _ Client = &draPluginClient{}
+
+type nodeV1ClientCreator func(addr draAddr) (
+	nodeClient drapbv1.NodeClient,
+	closer io.Closer,
+	err error,
+)
+
+// UncertainProgressError indicates operation failed with a non-final error
+// and operation may be in-progress in background.
+type UncertainProgressError struct {
+	msg string
+}
+
+func (err *UncertainProgressError) Error() string {
+	return err.msg
+}
+
+// NewUncertainProgressError creates an instance of UncertainProgressError type.
+func NewUncertainProgressError(msg string) *UncertainProgressError {
+	return &UncertainProgressError{msg: msg}
+}
```

Comment:
I'm not opposed to adding this error type like this, but I'm trying to understand what the advantages of having it are. Where / how is it checked to see if the error returned is a `UncertainProgressError` (and what do we do differently in response to this as opposed to any other error)?

Reply:
I tend to remove it as it looks like a leftover from csi code.

Reply:
after its removal `isFinalError` also doesn't look reasonable to keep in the code, it seems.